blog/_posts/2020-08-12-building-root-with-distcc.md

```diff
@@ -0,0 +1,85 @@
+---
+title: "Building ROOT fast - distributed compilation with distcc"
+layout: archive
+author: Stefan Wunsch
+---
+
+Hello there, here is Stefan with a short guide on how you can compile ROOT distributed with distcc!
+
+ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers:
+
+```
+Building ROOT on 4 cores: 86 minutes
+```
+
+How to make this faster? Besides buying a more powerful computer, another answer from the software community is [`distcc`](https://distcc.github.io/). `distcc` allows to compile C, C++ and objective C(++) distributed, is simple to set up and already integrated into the ROOT build system!
+
+## Setup
+
+But first we have to make another machine accessible for the distributed compilation, called a *volunteer* in the `distcc` terminology. Note that `distcc` distributes only the compilation step, not the preprocessing (without the [`pump` mode](https://distcc.github.io/faq.html)) and linking. This means that the volunteer does not need any dependencies installed, making the setup simple!
+
+```bash
+# Set up distcc on a second machine (a volunteer) using gcc as compiler
+# Take care that gcc has the same version than on the main system!
+sudo apt install gcc distcc # Or use your preferred package manager!
+
+# Run the distcc(d) daemon!
+distccd --daemon --allow-private
+```
+
+Now, let's set up your environment on the main system to detect the volunteer!
+
+```bash
+# Tell distcc where to find other servers and how many cores from the volunteers
+# should be used (for me it's the ip 192.168.1.225 and 12 cores)
+# Note that distcc communicates by default via the port 3632, which may be blocked
+# by your firewall!
+export DISTCC_HOSTS="localhost/4 192.168.1.225/12"
+
+# Clone ROOT (assuming all dependencies are installed)
+git clone https://www.github.com/root-project/root
+cd root/
+
+# Configure ROOT to use distcc
+# You can also use distcc together with `-Dccache=ON` to enable caching, which is
+# checked before the compilations are distributed.
+cmake . -DCMAKE_CXX_COMPILER="g++" -DCMAKE_C_COMPILER="gcc" -Ddistcc="ON"
+
+# Run the compilation distributed on 4+12 cores!
+make -j16
+```
+
+## Docker
+
+Most likely you want to make use of an existing system as a (part time) volunteer, which means that installing exactly the same compiler than on the main system can be non trivial. Luckily, [Docker](https://www.docker.com/) makes deploying services such as `distccd` simple and you can even use a Windows machine as volunteer! Just create the following [`Dockerfile`](https://docs.docker.com/engine/reference/builder/) and follow the instructions below.
+
+```dockerfile
+FROM ubuntu:focal
+RUN apt update -y && apt install -y gcc distcc htop && useradd distcc
+ENTRYPOINT distccd --allow-private --daemon && htop
+```
```

Comment:
might be worth specifying that this is if your first machine is ubuntu focal (I think?)

Reply:
well, the idea is that you just pick a system that has the compiler you need. it does not have to match your distro/OS. but at the end you'll pick the same one than you run because it's the trivial solution to the problem!

Reply:
sure, still I think it would be good to specify that this is not a dockerfile that will work for everyone: you need to pick a `FROM` with the same (or a similar enough) platform as your main machine.